doc/rst/language/evolution.rst

```diff
@@ -3,18 +3,225 @@
+be fewer adjustments of +/- 1 in the code.
+
+For these reasons, we decided to bite the bullet and make the switch
+now, while we felt we still could, rather than later when it would
+clearly be too late to do so and cause more of a revolt among our
+users.
+
+Index-neutral Features
+**********************
+
+This experience also led to a number of new programming features in
+Chapel 1.21 designed to help write code in more of an index-neutral
+style.  Chief among these are new `.indices` queries on most of the
+relevent types as well as support for loops over heterogeneous tuples.
+We also introduced features that we found to be useful in updating
+code, such as support for open-interval ranges and `.first` and
+`.last` queries on enumerated types.  To this end, even though Chapel
+still has cases that require making this 0- vs. 1-based indexing
+decision, we encourage code to be written in an index-neutral style
+whenever possible, and believe that most common code patterns can be.
+
+Tips for Updating Existing Chapel code
+**************************************
+
+The following are some tips for updating codes based on our
+experiences:
+
+* First, updating code is easiest when it has some sort of testing
+  infrastructure that can be used to validate that its behavior is
+  unchanged.  If you don't already have such testing for your code, it
+  may be worthwhile to invest in creating some before attempting this
+  upgrade.
+
+* Next, when transitioning code to Chapel 1.22, make sure to compile
+  it with neither `--fast` nor `--no-checks` enabled so that bounds
+  checks are turned on in the generated code.  In cases where a
+  program is accessing all of the elements of a collection (as is
+  common for tuples) this will help identify data structures that
+  require updates.  When you do get an out-of-bounds error, don't
+  simply update the specific access, but use it as a cue to look
+  through the code for other references to that variable that will
+  also need updating.
+
+* When possible, try rewriting your updated code to use an
+  index-neutral style of programming.  For example, given code like
+  this:
+
+  .. code-block:: chapel
+
+      var t: 2*int = ...;
+
+      var x = t(1),
+          y = t(2);
+
+      for i in 1..2 do
+        writeln("t(", i, ") = ", t(i));
+
+  It would be reasonable to rewrite it like this:
+
+  .. code-block:: chapel
+
+      var t: 2*int = ...;
+
+      var x = t(0),
+          y = t(1);
+
+      for i in 0..1 do
+        writeln("t(", i, ") = ", t(i));
+
+  But arguably preferable to update it like this:
+
+  .. code-block:: chapel
+
+      var t: 2*int = ...;
+
+      var (x, y) = t;
+
+      for i in t.indices do
+        writeln("t(", i, ") = ", t(i));
+
+  If you have a pattern that you're trying to write in an
+  index-neutral style, but can't, don't hesitate to ask for tips.
+        
+
+* Some common pitfalls to check for in your code include:
```

Comment:
Some others that may or may not be worth mentioning:

- Anything that is generating or modifying indices, such as `RandomStream.getNext()` or `ceil()` / `floor()`

- Some common strings to look for are:
  - `[1]`
  - `[2]`
  - `(1)`
  - `(2)`
  - `1..`

Reply:
I like the sentiment of thinking about things that compute indices, but am not sure that getNext, ceil, or floor are necessarily great indicators of such cases (vs. doing some other sort of computation including casting).  So what if we were to say something "Think about whether there are places in your code that compute index values numerically." as a final bullet?

(I like the "common strings" cases, though, and should've thought to put something like that).

Reply:
> Think about whether there are places in your code that compute index values numerically." as a final bullet?

That works.